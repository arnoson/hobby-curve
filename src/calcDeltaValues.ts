import { Knot } from './utils'

export const calcDeltaValues = (knots: Knot[], cyclic: boolean) => {
  const end = cyclic ? knots.length : knots.length - 1
  for (let i = 0; i < end; i++) {
    const knot = knots[i]
    const nextKnot = knots[i + 1] ?? knots[0]

    knot.deltaX = nextKnot.x - knot.x
    knot.deltaY = nextKnot.y - knot.y
    knot.delta = Math.hypot(knot.deltaX, knot.deltaY)
  }
}
