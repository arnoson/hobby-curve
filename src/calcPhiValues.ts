import { Knot } from './utils'

export const calcPhiValues = (knots: Knot[], cyclic) => {
  for (let i = 0; i < knots.length; i++) {
    const knot = knots[i]
    knot.phi = -knot.psi - knot.theta
  }
}
