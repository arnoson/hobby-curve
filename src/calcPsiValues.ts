import { Knot } from './utils'

export const calcPsiValues = (knots: Knot[], cyclic: boolean) => {
  const [start, end] = cyclic ? [0, knots.length] : [1, knots.length - 1]
  for (let i = start; i < end; i++) {
    const knot = knots[i]
    const prevKnot = knot.prev

    const sin = prevKnot.deltaY / prevKnot.delta
    const cos = prevKnot.deltaX / prevKnot.delta

    knot.psi = Math.atan2(
      knot.deltaY * cos - knot.deltaX * sin,
      knot.deltaX * cos + knot.deltaY * sin
    )
  }
}
