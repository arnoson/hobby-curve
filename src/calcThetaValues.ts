import { Knot, curlRatio } from './utils'

export const calcThetaValues = function (knots: Knot[], cyclic: boolean) {
  // The 'matrix' is in tridiagonal form, the solution is obtained by Gaussian
  // elimination. `uu` and `ww` are of type fraction, vv and theta are of type
  // angle.
  // Relations between adjacent angles ('matrix' entries).
  const uu = []
  // Additional matrix entries for the cyclic case.
  const ww = []
  // Angles ('rhs' entries)
  const vv = []

  const firstKnot = knots[0]
  const secondKnot = knots[1]
  const lastKnot = knots[knots.length - 1]
  // Cycle back to the first knot on cyclic paths.
  const passes = cyclic ? knots.length + 1 : knots.length

  // First pass
  if (!cyclic) {
    const nextKnot = firstKnot.next
    const gamma = firstKnot.rightX
    const leftTension = Math.abs(nextKnot.leftY)
    const rightTension = Math.abs(firstKnot.rightY)
    uu[0] = curlRatio(gamma, rightTension, leftTension)
    vv[0] = -(secondKnot.psi * uu[0])
    ww[0] = 0
  } else {
    uu[0] = 0
    vv[0] = 0
    ww[0] = 1
  }

  // Remaining passes
  for (let i = 1; i < passes; i++) {
    const knot = knots[i] ?? firstKnot
    const nextKnot = knot.next
    const prevKnot = knot.prev
    const isLast = i === passes - 1

    if (cyclic || !isLast) {
      // Set up equation to match mock curvatures at z_k; then finish loop with
      // theta_n adjusted to equal theta_0, if a cycle has ended.

      // Calculate the values:
      // aa = Ak/Bk, bb = Dk/Ck, dd = (3-alpha_{k-1})d(k,k+1),
      // ee = (3-beta_{k+1})d(k-1,k), cc=(Bk-uk-Ak)/Bk
      let aa = 1 / (3 * Math.abs(prevKnot!.rightY) - 1)
      let dd = knot.delta * (3 - 1 / Math.abs(prevKnot.rightY))

      let bb = 1 / (3 * Math.abs(nextKnot.leftY) - 1)
      let ee = prevKnot.delta * (3 - 1 / Math.abs(nextKnot.leftY))

      const cc = 1 - uu[i - 1] * aa

      // Calculate the ratio ff = Ck/(Ck + Bk - uk-1Ak).
      dd = dd * cc
      const tensionLeft = Math.abs(knot.leftY)
      const tensionRight = Math.abs(knot.rightY)
      if (tensionLeft < tensionRight) {
        dd *= (tensionLeft / tensionRight) ** 2
      } else {
        ee *= (tensionRight / tensionLeft) ** 2
      }
      let ff = ee / (ee + dd)
      uu[i] = ff * bb

      // Calculate the values of vk and wk.
      let acc = -(nextKnot.psi * uu[i])
      ff = (1 - ff) / cc
      acc = acc - knot.psi * ff
      ff = ff * aa
      vv[i] = acc - vv[i - 1] * ff
      ww[i] = -(ww[i - 1] * ff)

      // Adjust theta_n to equal theta_0 and finish loop.
      if (cyclic && isLast) {
        let aa = 0
        let bb = 1

        for (let j = i - 1; j >= 0; j--) {
          const index = j === 0 ? passes - 1 : j
          aa = vv[index] - aa * uu[index]
          bb = ww[index] - bb * uu[index]
        }

        aa = aa / (1 - bb)
        firstKnot.theta = aa
        vv[0] = aa
        for (let i = 1; i < passes - 1; i++) {
          vv[i] = vv[i] + aa * ww[i]
        }

        break
      }
    } else {
      // Last knot on a non-cyclic path.
      const gamma = knot.leftX
      const tensionLeft = Math.abs(knot.leftY)
      const tensionRight = Math.abs(prevKnot.rightY)
      const ff = curlRatio(gamma, tensionLeft, tensionRight)
      lastKnot.theta = -((vv[passes - 2] * ff) / (1 - ff * uu[passes - 2]))
      break
    }
  }

  for (let i = passes - 2; i >= 0; i -= 1) {
    knots[i].theta = vv[i] - knots[i].next.theta * uu[i]
  }
}
