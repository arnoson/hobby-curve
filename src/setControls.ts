import { Type, Knot, velocity, abGreaterCd } from './utils'

export const setControls = (
  knot: Knot,
  st: number,
  ct: number,
  sf: number,
  cf: number
) => {
  const nextKnot = knot.next

  let lt = Math.abs(nextKnot.leftY)
  let rt = Math.abs(knot.rightY)
  let rr = velocity(st, ct, sf, cf, rt)
  let ss = velocity(sf, cf, st, ct, lt)

  if (knot.rightY < 0 || nextKnot.leftY < 0) {
    // Decrease the velocities, if necessary, to stay inside the bounding
    // triangle.
    if ((st >= 0 && sf >= 0) || (st <= 0 && sf <= 0)) {
      let sine = Math.abs(st) * cf + Math.abs(sf) * ct
      if (sine > 0) {
        sine *= 1.00024414062 // safety factor
        if (knot.rightY < 0) {
          if (!abGreaterCd(Math.abs(sf), 1, rr, sine)) {
            rr = Math.abs(sf) / sine
          }
        }
        if (nextKnot.leftY < 0) {
          if (!abGreaterCd(Math.abs(st), 1, ss, sine)) {
            ss = Math.abs(st) / sine
          }
        }
      }
    }
  }

  knot.rightX = knot.x + (knot.deltaX * ct - knot.deltaY * st) * rr
  knot.rightY = knot.y + (knot.deltaY * ct + knot.deltaX * st) * rr
  knot.rightType = Type.Explicit

  nextKnot.leftX = nextKnot.x - (knot.deltaX * cf + knot.deltaY * sf) * ss
  nextKnot.leftY = nextKnot.y - (knot.deltaY * cf - knot.deltaX * sf) * ss
  nextKnot.leftType = Type.Explicit
}
