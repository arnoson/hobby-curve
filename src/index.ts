/**
 * The mathematics of this code is based on vlad-x's javascript port.
 * (https://github.com/vlad-x/hobby-curves) of Michael Schindler's Python
 * Implementation of some internal functions of MetaPost, developed by John
 * D. Hobby and others.
 * Quoted by Michael Schindler: "Metapost's code is in the public domain,
 * which we take as implicit permission to reuse the code here.
 * (see the comment at http://www.gnu.org/licenses/license-list.html)."
 *
 * I heavily refactored and simplified the code from vlad-x, ported it to
 * Typescript and was inspired by Luke Trujillo's javascript implementation
 * (https://github.com/ltrujello/Hobby_Curve_Algorithm/tree/main/javascript)
 * especially for naming and structuring things.
 *
 * I also added and modified some of the comments from Michael Schindler's
 * python implementation to get a better understanding.
 */

import { Point } from './utils'
import { calcDeltaValues } from './calcDeltaValues'
import { calcPsiValues } from './calcPsiValues'
import { calcPhiValues } from './calcPhiValues'
import { calcThetaValues } from './calcThetaValues'
import { setControls, setControlsLine } from './setControls'

const createKnot = (x: number, y: number, tension: number) => ({
  x: x,
  y: y,

  leftY: tension,
  rightY: tension,

  leftX: tension,
  rightX: tension,

  deltaX: 0,
  deltaY: 0,
  delta: 0,

  theta: 0,
  phi: 0,
  psi: 0,
})

const createHobbyKnots = (
  points: Point[],
  tension: any = 1,
  cyclic = false
) => {
  // @ts-ignore (`next` und `prev` are missing, but will be set immediately)
  const knots: Knot[] = points.map(({ x, y }) => createKnot(x, y, tension))
  const firstKnot = knots[0]
  const lastKnot = knots[knots.length - 1]
  for (let i = 0; i < knots.length; i++) {
    knots[i].next = knots[i + 1] ?? firstKnot
    knots[i].prev = knots[i - 1] ?? lastKnot
  }

  calcDeltaValues(knots, cyclic)
  // If we only have two points on a non-cyclic path we can take a shortcut and
  // just set the control points as a straight line.
  if (points.length === 2 && !cyclic) {
    setControlsLine(firstKnot, lastKnot)
    return knots
  }
  calcPsiValues(knots, cyclic)
  calcThetaValues(knots, cyclic)
  calcPhiValues(knots, cyclic)

  const end = cyclic ? knots.length : knots.length - 1
  for (let i = 0; i < end; i++) {
    setControls(knots[i], knots[i].next)
  }

  return knots
}

/**
 * Create a hobby curve and return a list of bezier entries in the format
 * `{ startControl, endControl, point }`.
 */
export const createHobbyBezier = (
  points: Point[],
  { tension = 1, cyclic = false } = {}
) => {
  const knots = createHobbyKnots(points, tension, cyclic)
  const bezier: { startControl: Point; endControl: Point; point: Point }[] = []

  const end = cyclic ? knots.length : knots.length - 1
  for (let i = 0; i < end; i++) {
    const knot = knots[i]
    bezier.push({
      startControl: { x: knot.rightX, y: knot.rightY },
      endControl: { x: knot.next.leftX, y: knot.next.leftY },
      point: { x: knot.next.x, y: knot.next.y },
    })
  }

  return bezier
}

/**
 * Create a hobby curve and return it's path definition.
 */
export const createHobbyCurve = (
  points: Point[],
  { tension = 1, cyclic = false } = {}
) => {
  const bezier = createHobbyBezier(points, { tension, cyclic })

  const toString = (point: Point) => [point.x, point.y].join(',')
  const bezierCommands = bezier.map(({ startControl, endControl, point }) =>
    [toString(startControl), toString(endControl), toString(point)].join(' ')
  )

  return `M ${points[0].x},${points[0].y} C ${bezierCommands}`
}
