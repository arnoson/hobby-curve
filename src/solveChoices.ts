import { Knot, Type, curlRatio, sinCos } from './utils'
import { setControls } from './setControls'

const UNITY = 1

export const solveChoices = function (knots: Knot[], cyclic: boolean) {
  // The 'matrix' is in tridiagonal form, the solution is obtained by Gaussian
  // elimination. `uu` and `ww` are of type fraction, vv and theta are of type
  // angle.
  // Relations between adjacent angles ('matrix' entries).
  const uu = []
  // Additional matrix entries for the cyclic case.
  const ww = []
  // Angles ('rhs' entries)
  const vv = []
  // Solution of the linear system of equations.
  const theta = []

  const firstKnot = knots[0]
  const secondKnot = knots[1]
  const passes = cyclic ? knots.length + 1 : knots.length

  // Solve first knot.
  if (firstKnot.rightType == Type.Curl) {
    const nextKnot = firstKnot.next

    if (nextKnot.leftType == Type.Curl) {
      // Reduce to simple case of straight line and return.
      firstKnot.rightType = Type.Explicit
      nextKnot.leftType = Type.Explicit
      const lt = Math.abs(nextKnot.leftY)
      const rt = Math.abs(firstKnot.rightY)

      let ff = UNITY / (3.0 * rt)
      firstKnot.rightX = firstKnot.x + firstKnot.deltaX * ff
      firstKnot.rightY = firstKnot.y + firstKnot.deltaY * ff

      ff = UNITY / (3.0 * lt)
      nextKnot.leftX = nextKnot.x - firstKnot.deltaX * ff
      nextKnot.leftY = nextKnot.y - firstKnot.deltaY * ff

      return
    } else {
      const cc = firstKnot.rightX
      const lt = Math.abs(nextKnot.leftY)
      const rt = Math.abs(firstKnot.rightY)
      uu[0] = curlRatio(cc, rt, lt)
      vv[0] = -(secondKnot.psi * uu[0])
      ww[0] = 0
    }
  } else {
    if (firstKnot.rightType == Type.Open) {
      uu[0] = 0
      vv[0] = 0
      ww[0] = 1
    }
  }

  // Solve remaining knots (and cycle back to first if path is cyclic).
  for (let i = 1; i < passes; i++) {
    const knot = knots[i] ?? firstKnot
    const nextKnot = knot.next
    const prevKnot = knot.prev
    const isLast = i === passes - 1

    if (cyclic || !isLast) {
      // Set up equation to match mock curvatures at z_k; then finish loop with
      // theta_n adjusted to equal theta_0, if a cycle has ended.

      // Calculate the values:
      // aa = Ak/Bk, bb = Dk/Ck, dd = (3-alpha_{k-1})d(k,k+1),
      // ee = (3-beta_{k+1})d(k-1,k), cc=(Bk-uk-Ak)/Bk
      let aa = UNITY / (3.0 * Math.abs(prevKnot!.rightY) - UNITY)
      let dd = knot.delta * (3 - UNITY / Math.abs(prevKnot!.rightY))

      let bb = UNITY / (3 * Math.abs(nextKnot.leftY) - UNITY)
      let ee = prevKnot.delta * (3 - UNITY / Math.abs(nextKnot.leftY))

      const cc = 1 - uu[i - 1] * aa

      // Calculate the ratio ff = Ck/(Ck + Bk - uk-1Ak).
      dd = dd * cc
      const lt = Math.abs(knot.leftY)
      const rt = Math.abs(knot.rightY)
      if (lt < rt) {
        dd *= (lt / rt) ** 2
      } else {
        if (lt > rt) {
          ee *= (rt / lt) ** 2
        }
      }
      let ff = ee / (ee + dd)
      uu[i] = ff * bb

      // Calculate the values of vk and wk
      let acc = -(nextKnot.psi * uu[i])
      if (prevKnot!.rightType == Type.Curl) {
        ww[i] = 0
        vv[i] = acc - secondKnot.psi * (1 - ff)
      } else {
        ff = (1 - ff) / cc
        acc = acc - knot.psi * ff
        ff = ff * aa
        vv[i] = acc - vv[i - 1] * ff
        ww[i] = -(ww[i - 1] * ff)
      }

      // Adjust theta_n to equal theta_0 and finish loop.
      if (cyclic && isLast) {
        let aa = 0
        let bb = 1

        for (let j = i - 1; j >= 0; j--) {
          const index = j === 0 ? passes - 1 : j
          aa = vv[index] - aa * uu[index]
          bb = ww[index] - bb * uu[index]
        }

        aa = aa / (1 - bb)
        theta[passes - 1] = aa
        vv[0] = aa
        for (let i = 1; i < passes - 1; i++) {
          vv[i] = vv[i] + aa * ww[i]
        }

        break
      }
    } else {
      // Last knot on a non-cyclic path.
      const cc = knot.leftX
      const lt = Math.abs(knot.leftY)
      const rt = Math.abs(prevKnot!.rightY)
      const ff = curlRatio(cc, lt, rt)
      theta[passes - 1] = -((vv[passes - 2] * ff) / (1 - ff * uu[passes - 2]))
      break
    }
  }

  // Finish choosing angles and assigning control points.
  for (let i = passes - 2; i > -1; i -= 1) {
    theta[i] = vv[i] - theta[i + 1] * uu[i]
  }

  for (let i = 0; i < passes - 1; i++) {
    const knot = knots[i] ?? firstKnot
    const nextKnot = knot.next
    const [ct, st] = sinCos(theta[i])
    const [cf, sf] = sinCos(-nextKnot.psi - theta[i + 1])
    setControls(knot, st, ct, sf, cf)
  }
}
