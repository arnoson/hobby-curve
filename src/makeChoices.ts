import { Knot } from './utils'
import { solveChoices } from './solveChoices'

const calcDeltaValues = (knots: Knot[], cyclic: boolean) => {
  const end = cyclic ? knots.length : knots.length - 1
  for (let i = 0; i < end; i++) {
    const knot = knots[i]
    const nextKnot = knots[i + 1] ?? knots[0]

    knot.deltaX = nextKnot.x - knot.x
    knot.deltaY = nextKnot.y - knot.y
    knot.delta = Math.hypot(knot.deltaX, knot.deltaY)
  }
}

const calcPsiValues = (knots: Knot[], cyclic: boolean) => {
  const [start, end] = cyclic ? [0, knots.length] : [1, knots.length - 1]
  for (let i = start; i < end; i++) {
    const knot = knots[i]
    const prevKnot = knot.prev

    const sin = prevKnot.deltaY / prevKnot.delta
    const cos = prevKnot.deltaX / prevKnot.delta

    knot.psi = Math.atan2(
      knot.deltaY * cos - knot.deltaX * sin,
      knot.deltaX * cos + knot.deltaY * sin
    )
  }
}

export const makeChoices = (knots: Knot[], cyclic = true) => {
  calcDeltaValues(knots, cyclic)
  calcPsiValues(knots, cyclic)
  solveChoices(knots, cyclic)
}
