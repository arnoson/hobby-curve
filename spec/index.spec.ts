import { createHobbyBezier, createHobbyCurve } from '../src/index'

const points = [
  { x: 0, y: 0 },
  { x: 200, y: 133 },
  { x: 130, y: 300 },
  { x: 33, y: 233 },
  { x: 100, y: 167 },
]

describe('createHobbyBezier', () => {
  it('handles cyclic and non-cyclic paths', () => {
    expect(createHobbyBezier(points, 1, true)).toMatchSnapshot()
    expect(createHobbyBezier(points, 1, false)).toMatchSnapshot()
  })

  it('handles 2-point paths', () => {
    const points = [
      { x: 0, y: 0 },
      { x: 200, y: 133 },
    ]
    expect(createHobbyBezier(points, 1, true)).toMatchSnapshot()
    expect(createHobbyBezier(points, 1, false)).toMatchSnapshot()
  })

  it('handles low tension', () => {
    expect(createHobbyBezier(points, 0.3, true)).toMatchSnapshot()
    expect(createHobbyBezier(points, 0.3, false)).toMatchSnapshot()
  })

  it('handles high tension', () => {
    expect(createHobbyBezier(points, 4, true)).toMatchSnapshot()
    expect(createHobbyBezier(points, 4, false)).toMatchSnapshot()
  })
})

describe('createHobbyCurve', () => {
  it('creates an svg path description', () => {
    expect(createHobbyCurve(points, 1, true)).toMatchSnapshot()
    expect(createHobbyCurve(points, 1, false)).toMatchSnapshot()
  })
})
