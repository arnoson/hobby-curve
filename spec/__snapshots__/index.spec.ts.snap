// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createHobbyBezier handles 2-point paths 1`] = `
Array [
  Object {
    "endControl": Object {
      "x": 288.66666666666663,
      "y": -0.3333333333333144,
    },
    "point": Object {
      "x": 200,
      "y": 133,
    },
    "startControl": Object {
      "x": 88.66666666666666,
      "y": -133.33333333333331,
    },
  },
  Object {
    "endControl": Object {
      "x": -88.66666666666666,
      "y": 133.33333333333331,
    },
    "point": Object {
      "x": 0,
      "y": 0,
    },
    "startControl": Object {
      "x": 111.33333333333334,
      "y": 266.3333333333333,
    },
  },
]
`;

exports[`createHobbyBezier handles 2-point paths 2`] = `
Array [
  Object {
    "endControl": Object {
      "x": 133.33333333333334,
      "y": 88.66666666666667,
    },
    "point": Object {
      "x": 200,
      "y": 133,
    },
    "startControl": Object {
      "x": 66.66666666666666,
      "y": 44.33333333333333,
    },
  },
]
`;

exports[`createHobbyBezier handles cyclic and non-cyclic paths 1`] = `
Array [
  Object {
    "endControl": Object {
      "x": 202.2726597318875,
      "y": -62.08929209418508,
    },
    "point": Object {
      "x": 200,
      "y": 133,
    },
    "startControl": Object {
      "x": 17.127431909906143,
      "y": -89.74523875089386,
    },
  },
  Object {
    "endControl": Object {
      "x": 189.20267180945046,
      "y": 273.5595327118616,
    },
    "point": Object {
      "x": 130,
      "y": 300,
    },
    "startControl": Object {
      "x": 199.21702504025157,
      "y": 200.2120108793997,
    },
  },
  Object {
    "endControl": Object {
      "x": 15.74830065248955,
      "y": 279.98057448405984,
    },
    "point": Object {
      "x": 33,
      "y": 233,
    },
    "startControl": Object {
      "x": 72.32351951032112,
      "y": 325.75885596161345,
    },
  },
  Object {
    "endControl": Object {
      "x": 87.60189583018445,
      "y": 197.254578946683,
    },
    "point": Object {
      "x": 100,
      "y": 167,
    },
    "startControl": Object {
      "x": 44.337677624281866,
      "y": 202.12474548887275,
    },
  },
  Object {
    "endControl": Object {
      "x": -13.489254929089425,
      "y": 70.68172336347855,
    },
    "point": Object {
      "x": 0,
      "y": 0,
    },
    "startControl": Object {
      "x": 131.5922197261313,
      "y": 89.90681763005249,
    },
  },
]
`;

exports[`createHobbyBezier handles cyclic and non-cyclic paths 2`] = `
Array [
  Object {
    "endControl": Object {
      "x": 171.7811681131284,
      "y": 48.00234580207959,
    },
    "point": Object {
      "x": 200,
      "y": 133,
    },
    "startControl": Object {
      "x": 89.29718943918873,
      "y": -6.849500016090294,
    },
  },
  Object {
    "endControl": Object {
      "x": 197.16569405952507,
      "y": 283.6273089083985,
    },
    "point": Object {
      "x": 130,
      "y": 300,
    },
    "startControl": Object {
      "x": 223.64739995026858,
      "y": 204.22809093270655,
    },
  },
  Object {
    "endControl": Object {
      "x": 34.380121098298794,
      "y": 280.179075007001,
    },
    "point": Object {
      "x": 33,
      "y": 233,
    },
    "startControl": Object {
      "x": 82.08937210171507,
      "y": 311.6789667934957,
    },
  },
  Object {
    "endControl": Object {
      "x": 62.73018959428872,
      "y": 165.34854639397165,
    },
    "point": Object {
      "x": 100,
      "y": 167,
    },
    "startControl": Object {
      "x": 31.909149699518927,
      "y": 195.709570767924,
    },
  },
]
`;

exports[`createHobbyBezier handles high tension 1`] = `
Array [
  Object {
    "endControl": Object {
      "x": 193.13386237171488,
      "y": 97.17379981667526,
    },
    "point": Object {
      "x": 200,
      "y": 133,
    },
    "startControl": Object {
      "x": 12.157922585803924,
      "y": -18.16527386600955,
    },
  },
  Object {
    "endControl": Object {
      "x": 146.58893158115984,
      "y": 297.78978588457517,
    },
    "point": Object {
      "x": 130,
      "y": 300,
    },
    "startControl": Object {
      "x": 203.6456020166292,
      "y": 152.02205791775114,
    },
  },
  Object {
    "endControl": Object {
      "x": 32.28550270549024,
      "y": 244.26217321503026,
    },
    "point": Object {
      "x": 33,
      "y": 233,
    },
    "startControl": Object {
      "x": 117.28176710390673,
      "y": 301.69450441896635,
    },
  },
  Object {
    "endControl": Object {
      "x": 97.0453960963215,
      "y": 174.61489244465952,
    },
    "point": Object {
      "x": 100,
      "y": 167,
    },
    "startControl": Object {
      "x": 33.56879807608968,
      "y": 224.03439228318814,
    },
  },
  Object {
    "endControl": Object {
      "x": -13.323517723023173,
      "y": 19.906801230988783,
    },
    "point": Object {
      "x": 0,
      "y": 0,
    },
    "startControl": Object {
      "x": 106.79682646699854,
      "y": 149.48255800150702,
    },
  },
]
`;

exports[`createHobbyBezier handles high tension 2`] = `
Array [
  Object {
    "endControl": Object {
      "x": 192.25232442538717,
      "y": 82.97833889609893,
    },
    "point": Object {
      "x": 200,
      "y": 133,
    },
    "startControl": Object {
      "x": -10.128850571697496,
      "y": -29.264839684712893,
    },
  },
  Object {
    "endControl": Object {
      "x": 146.53601039740755,
      "y": 297.91976580586646,
    },
    "point": Object {
      "x": 130,
      "y": 300,
    },
    "startControl": Object {
      "x": 202.95316078090127,
      "y": 152.0666228012467,
    },
  },
  Object {
    "endControl": Object {
      "x": 34.04574657244422,
      "y": 244.2531062479157,
    },
    "point": Object {
      "x": 33,
      "y": 233,
    },
    "startControl": Object {
      "x": 118.19764809062058,
      "y": 301.4847387866264,
    },
  },
  Object {
    "endControl": Object {
      "x": 104.55506630339977,
      "y": 152.68468266701754,
    },
    "point": Object {
      "x": 100,
      "y": 167,
    },
    "startControl": Object {
      "x": 31.162885786451195,
      "y": 213.23111719477004,
    },
  },
]
`;

exports[`createHobbyBezier handles low tension 1`] = `
Array [
  Object {
    "endControl": Object {
      "x": -97.58399849640273,
      "y": 1046.4920710323067,
    },
    "point": Object {
      "x": 200,
      "y": 133,
    },
    "startControl": Object {
      "x": -958.4026556309514,
      "y": 66.99514668645821,
    },
  },
  Object {
    "endControl": Object {
      "x": 373.2193448539864,
      "y": 982.2524095148347,
    },
    "point": Object {
      "x": 130,
      "y": 300,
    },
    "startControl": Object {
      "x": 424.3505586762873,
      "y": -555.6877571306446,
    },
  },
  Object {
    "endControl": Object {
      "x": -401.1829761298074,
      "y": 329.9248157967111,
    },
    "point": Object {
      "x": 33,
      "y": 233,
    },
    "startControl": Object {
      "x": 95.42060754055463,
      "y": 203.00165540212777,
    },
  },
  Object {
    "endControl": Object {
      "x": 43.93592731724532,
      "y": 261.40077959610556,
    },
    "point": Object {
      "x": 100,
      "y": 167,
    },
    "startControl": Object {
      "x": 139.68694685728406,
      "y": 209.18374454883224,
    },
  },
  Object {
    "endControl": Object {
      "x": 292.0205413561221,
      "y": -20.413089309245105,
    },
    "point": Object {
      "x": 0,
      "y": 0,
    },
    "startControl": Object {
      "x": 252.26184849743314,
      "y": -89.37875582526374,
    },
  },
]
`;

exports[`createHobbyBezier handles low tension 2`] = `
Array [
  Object {
    "endControl": Object {
      "x": -75.51395986470311,
      "y": 222.8782686605149,
    },
    "point": Object {
      "x": 200,
      "y": 133,
    },
    "startControl": Object {
      "x": -474.70993135830554,
      "y": -835.2691069767832,
    },
  },
  Object {
    "endControl": Object {
      "x": -230.8620846323861,
      "y": 928.0147736118703,
    },
    "point": Object {
      "x": 130,
      "y": 300,
    },
    "startControl": Object {
      "x": 888.5954047155858,
      "y": -91.633854537954,
    },
  },
  Object {
    "endControl": Object {
      "x": 51.098593650518225,
      "y": 348.14475910237087,
    },
    "point": Object {
      "x": 33,
      "y": 233,
    },
    "startControl": Object {
      "x": 264.19987758735533,
      "y": 66.44954033444631,
    },
  },
  Object {
    "endControl": Object {
      "x": 473.29449500776747,
      "y": 120.40364824478343,
    },
    "point": Object {
      "x": 100,
      "y": 167,
    },
    "startControl": Object {
      "x": 19.047842084232496,
      "y": 144.23521013892218,
    },
  },
]
`;

exports[`createHobbyCurve creates an svg path description 1`] = `"M 0,0 C 17.127431909906143,-89.74523875089386 202.2726597318875,-62.08929209418508 200,133,199.21702504025157,200.2120108793997 189.20267180945046,273.5595327118616 130,300,72.32351951032112,325.75885596161345 15.74830065248955,279.98057448405984 33,233,44.337677624281866,202.12474548887275 87.60189583018445,197.254578946683 100,167,131.5922197261313,89.90681763005249 -13.489254929089425,70.68172336347855 0,0"`;

exports[`createHobbyCurve creates an svg path description 2`] = `"M 0,0 C 89.29718943918873,-6.849500016090294 171.7811681131284,48.00234580207959 200,133,223.64739995026858,204.22809093270655 197.16569405952507,283.6273089083985 130,300,82.08937210171507,311.6789667934957 34.380121098298794,280.179075007001 33,233,31.909149699518927,195.709570767924 62.73018959428872,165.34854639397165 100,167"`;
